### DISCUSSION TITLE
`diesel::joinable!`'s not being added to `schema.rs` when there's a circular relation

### DISCUSSION THREAD

**Discussion by Levyks** [diesel-rs/diesel#3891] [Q&A] [answered]
Whenever I add a "circular" relation in my database migration, no `joinable` macros are added to my auto-generated schema file, however, if I comment out one of the two references, the other one remaining gets the `joinable` added as expected

```sql
CREATE TABLE rooms (id BIGINT PRIMARY KEY);

CREATE TABLE players (
    id BIGINT PRIMARY KEY,
    room_id BIGINT NOT NULL REFERENCES rooms(id)
);

ALTER TABLE rooms
    ADD COLUMN leader_id BIGINT NULL REFERENCES players(id);
```

## Versions

- Rust: 1.74.1
- Diesel CLI: 2.1.1
- Database:  PostgreSQL 16.1
- OS: Windows 11 22H2

## Reproduction
[Levyks/diesel-relation-bug-reproduction](https://github.com/Levyks/diesel-relation-bug-reproduction)

Opening a discussion with more details as [advised here](https://matrix.to/#/!lNGJpfiFVovXFJYmwx:matrix.org/$170378495541UpMPo:gitter.im?via=gitter.im&via=oxide.computer)

**Comment by weiznich** [comment, marked as answer]
Thanks for filling this discussion. I had another look at the provided example and it's now clear for me what happens here. In short: This is expected behavior.

`joinable!(table_a -> table_b (fk))` enables you to write `table_a.inner_join(table_b)` and `table_b.inner_join(table_a)` (also `.left_join()`). If you now would have a `joinable!` entry for both your relations that syntax wouldn't be unambiguous anymore. It wouldn't be clear which relations should be used in this cases. For this reason the CLI tool filters out that kind of relations and requires you to write an explicit `on` clause which specifies which relation should be used instead.
